akka-sample-distributed-data-scala/src/main/scala/sample/distributeddata/ShoppingORCart.scala

```diff
@@ -0,0 +1,143 @@
+import akka.actor.Actor
+import akka.actor.ActorRef
+import akka.actor.Props
+import akka.cluster.Cluster
+import akka.cluster.ddata.{DistributedData, ORMultiMap, ORMultiMapKey}
+
+import scala.collection.immutable.HashSet
+
+object ShoppingORCart {
+  import akka.cluster.ddata.Replicator._
+
+  def props(userId: String): Props = Props(new ShoppingORCart(userId))
+
+  case object GetCart
+  final case class ChangeItemQuantity(item: LineItem)
+  final case class RemoveItem(productId: String)
+
+  final case class Cart(items: Set[LineItem])
+  final case class LineItem(productId: String, title: String, quantity: Int)
+  final case class SingleLineItem(productId: String, title: String, uid: String)
+
+  //#read-write-majority-orcart
+  private val timeout = 3.seconds
+  private val readMajority = ReadMajority(timeout)
+  private val writeMajority = WriteMajority(timeout)
+  //#read-write-majority-orcart
+
+  def getUniqueId(implicit cluster: Cluster): String = {
+    cluster.selfUniqueAddress.toString + UUID.randomUUID().toString
+  }
+
+}
+
+class ShoppingORCart(userId: String) extends Actor {
+  import ShoppingORCart._
+  import akka.cluster.ddata.Replicator._
+
+  val replicator = DistributedData(context.system).replicator
+  implicit val cluster = Cluster(context.system)
+
+  val DataKey = ORMultiMapKey[String, SingleLineItem]("cart-" + userId)
+
+  def receive = receiveGetCart
+    .orElse[Any, Unit](receiveAddItem)
+    .orElse[Any, Unit](receiveRemoveItem)
+    .orElse[Any, Unit](receiveOther)
+
+  //#get-orcart
+  def receiveGetCart: Receive = {
+    case GetCart ⇒
+      replicator ! Get(DataKey, readMajority, Some(sender()))
+
+    case g @ GetSuccess(DataKey, Some(replyTo: ActorRef)) ⇒
+      val data = g.get(DataKey)
+      val entrySet = data.entries.groupBy(_._1).map {
+        entry ⇒ LineItem(entry._2.values.head.head.productId, entry._2.values.head.head.title, entry._2.values.head.size)
+      }.toSet
+      val cart = Cart(entrySet)
+      replyTo ! cart
+
+    case NotFound(DataKey, Some(replyTo: ActorRef)) ⇒
+      replyTo ! Cart(Set.empty)
+
+    case GetFailure(DataKey, Some(replyTo: ActorRef)) ⇒
+      // ReadMajority failure, try again with local read
+      replicator ! Get(DataKey, ReadLocal, Some(replyTo))
+  }
+  //#get-orcart
+
+  //#add-item-orcart
+  def receiveAddItem: Receive = {
+    case cmd @ ChangeItemQuantity(item) ⇒
+      val update = Update(DataKey, ORMultiMap.emptyWithValueDeltas[String, SingleLineItem], writeMajority, Some(cmd)) {
+        cart ⇒ updateCart(cart, item)
+      }
+      replicator ! update
+  }
+  //#add-item-orcart
+
+  def updateCart(data: ORMultiMap[String, SingleLineItem], item: LineItem): ORMultiMap[String, SingleLineItem] =
+    data.get(item.productId) match {
+      case Some(entries) ⇒
+        val existingQuantity = entries.size
+        if (existingQuantity == item.quantity) {
+          data
+        } else if (existingQuantity < item.quantity) {
```

Comment:
Hmm it's not quite obvious to me that this converges... can we make that more clear-cut?

Reply:
There is quite a bit of logic in this actor. Shouldn't that logic be moved to a custom `ReplicatedData` subclass so it can expose methods like `updateCart` itself, and can be tested in isolation?

Reply:
I agree, this isn't obvious how it works. Would be good with a description and some tests.

It should not be a subclass of `ReplicatedData` because then we would add the whole serialization complexity to this sample, but it can be a wrapper to encapsulate the logic.

Maybe we should let @gosubpl complete this, since it's not our top priority to spend a lot of time on this right now.